README.md
### Features

- Select the folder
- Select the quality
- Change all png, jpg, jpeg files in folder and subfolder into Webp with the same name.

# Webp folder converter

![](https://img.shields.io/github/stars/FrancoisSiderikoudis/Webp-Folder-Converter.svg) ![](https://img.shields.io/github/forks/FrancoisSiderikoudis/Webp-Folder-Converter.svg) ![](https://img.shields.io/github/tag/FrancoisSiderikoudis/Webp-Folder-Converter.svg) ![](https://img.shields.io/github/release/FrancoisSiderikoudis/Webp-Folder-Converter.svg) ![](https://img.shields.io/github/issues/FrancoisSiderikoudis/Webp-Folder-Converter.svg)

### Why this project

I made this personnal project to learn and use for a first time electron and react.
This is for educational purpose only.

### How start the project

##### Prerequisite

* Node Js
* Yarn

```
cd "the current repo"
yarn install
yarn start
```

### License
© MIT

[Siderikoudis](https://www.siderikoudis.fr/)

[Electron](https://www.electronjs.org/)
[React](https://reactjs.org/)
[React Material UI](https://material-ui.com/)
[Electron React Boilerplate](https://github.com/electron-react-boilerplate/electron-react-boilerplate)
[recursive-readdir](https://www.npmjs.com/package/recursive-readdir)
[webp-converter](https://www.npmjs.com/package/webp-converter)

CHANGELOG.md
# 0.0.4 (06.03.2021)

#### Features

- **Select Folder**


src/main.ts
// @ts-nocheck

import { app, BrowserWindow, Menu, MenuItem, MenuItemConstructorOptions } from "electron";
import * as path from "path";

const { dialog } = require('electron')

let mainWindow : any = undefined;

function openAboutWindow(file : string) {
  const win = new BrowserWindow({
    height: 300,
    webPreferences: {
      devTools : false,
      nodeIntegration: false,
    },
    width: 200,
    resizable : false,
    autoHideMenuBar : true,
  })
  win.loadFile(path.join(__dirname, file));
};

function createWindow() {
  mainWindow = new BrowserWindow({
    height: 400,
    webPreferences: {
      devTools : false,
      preload: path.join(__dirname, "preload.js"),
      nodeIntegration: true,
    },
    width: 560,
    resizable : false,
    autoHideMenuBar : false,
  });

  mainWindow.loadFile(path.join(__dirname, "../index.html"));

  const { app, Menu } = require('electron')

  const isMac = process.platform === 'darwin'
    
  let menuTemplate = [{label: "Aide",submenu: [{ label: "A propos",click() {openAboutWindow("../APropos.html")}},
                      { label: "Licence",click() {openAboutWindow("../Licence.html")}},]},];
  
  let menu = Menu.buildFromTemplate(menuTemplate);
  Menu.setApplicationMenu(menu);
}

app.on("ready", () => {
  createWindow();

  app.on("activate", function () {
    if (BrowserWindow.getAllWindows().length === 0) createWindow(); 
  });
});

app.on("window-all-closed", () => {
  if (process.platform !== "darwin") {
    app.quit();
  }
});

const { ipcMain } = require('electron');
var recursive = require("recursive-readdir");
const webp = require('webp-converter');
webp.grant_permission();
const fs = require('fs')

var FolderLocation : string = undefined;
let exclusionFolder = ["node_modules"];
let quality = 80;

ipcMain.on('ChooseFolder', () => {
  dialog.showOpenDialog({ properties: ['openDirectory'] }).then((resolve)=> {
    FolderLocation = resolve.filePaths[0];
    let hamdou : string = resolve.filePaths[0].replace(/\\/g, "/");    
    mainWindow.webContents.executeJavaScript(`
        document.getElementById("Selected").innerHTML = '${hamdou}'
    `)
  })
});

ipcMain.on('inputChanged', (channel, inputValue) => {
  quality = inputValue;
});

ipcMain.on('startWebpGen', () => {
  recursive(FolderLocation, [] , function (err : any, files : any) {
    let rst = files.filter(function(file : any) {
        if (path.extname(file).toLowerCase() === ".png") return true;
        if (path.extname(file).toLowerCase() === ".jpg") return true;
        if (path.extname(file).toLowerCase() === ".jpeg") return true;
        return false;
    });
    let i = 0;
    mainWindow.webContents.executeJavaScript(`
      document.getElementById("NbrFichierATransformerEnWebp").innerHTML = 'Génération des webp en cours ${i}' / '${rst.length}'
    `);
    rst.forEach((Element : any) => {
          const result = webp.cwebp(Element, path.dirname(Element) + "\\" + path.basename(Element, path.extname(Element))  + ".webp","-q " + quality);
          result.then((result : any) => {
          ++i;
          mainWindow.webContents.executeJavaScript(`
            document.getElementById("CurrentAction").innerHTML = 'Génération des webp en cours ${i} / ${rst.length}'
          `);
        });
      }); // Fin foreach
  }); // Fin recursive
});

src/main.dev.ts
/* eslint global-require: off, no-console: off */

/**
 * This module executes inside of electron's main process. You can start
 * electron renderer process from here and communicate with the other processes
 * through IPC.
 *
 * When running `yarn build` or `yarn build-main`, this file is compiled to
 * `./src/main.prod.js` using webpack. This gives us some performance wins.
 */
// @ts-nocheck

import 'core-js/stable';
import 'regenerator-runtime/runtime';
import path from 'path';
import { app, BrowserWindow, shell, dialog } from 'electron';
import { autoUpdater } from 'electron-updater';
import log from 'electron-log';
import MenuBuilder from './menu';

export default class AppUpdater {
  constructor() {
    log.transports.file.level = 'info';
    autoUpdater.logger = log;
    autoUpdater.checkForUpdatesAndNotify();
  }
}

let mainWindow: BrowserWindow | null = null;

if (process.env.NODE_ENV === 'production') {
  const sourceMapSupport = require('source-map-support');
  sourceMapSupport.install();
}

if (
  process.env.NODE_ENV === 'development' ||
  process.env.DEBUG_PROD === 'true'
) {
  require('electron-debug')();
}

const installExtensions = async () => {
  const installer = require('electron-devtools-installer');
  const forceDownload = !!process.env.UPGRADE_EXTENSIONS;
  const extensions = ['REACT_DEVELOPER_TOOLS'];

  return installer
    .default(
      extensions.map((name) => installer[name]),
      forceDownload
    )
    .catch(console.log);
};

const createWindow = async () => {
  if (
    process.env.NODE_ENV === 'development' ||
    process.env.DEBUG_PROD === 'true'
  ) {
    await installExtensions();
  }

  const RESOURCES_PATH = app.isPackaged
    ? path.join(process.resourcesPath, 'assets')
    : path.join(__dirname, '../assets');

  const getAssetPath = (...paths: string[]): string => {
    return path.join(RESOURCES_PATH, ...paths);
  };

  mainWindow = new BrowserWindow({
    show: false,
    width: 500,
    height: 400,
    autoHideMenuBar : true,
    icon: getAssetPath('icon.png'),
    webPreferences: {
      nodeIntegration: true,
      enableRemoteModule: true,
    },
    resizable : false,
    frame:false,
  });

  mainWindow.loadURL(`file://${__dirname}/index.html`);

  // @TODO: Use 'ready-to-show' event
  //        https://github.com/electron/electron/blob/master/docs/api/browser-window.md#using-ready-to-show-event
  mainWindow.webContents.on('did-finish-load', () => {
    if (!mainWindow) {
      throw new Error('"mainWindow" is not defined');
    }
    if (process.env.START_MINIMIZED) {
      mainWindow.minimize();
    } else {
      mainWindow.show();
      mainWindow.focus();
    }
  });

  mainWindow.on('closed', () => {
    mainWindow = null;
  });

  const menuBuilder = new MenuBuilder(mainWindow);
  menuBuilder.buildMenu();

  // Open urls in the user's browser
  mainWindow.webContents.on('new-window', (event, url) => {
    event.preventDefault();
    shell.openExternal(url);
  });

  // Remove this if your app does not use auto updates
  // eslint-disable-next-line
  new AppUpdater();
};

/**
 * Add event listeners...
 */

app.on('window-all-closed', () => {
  // Respect the OSX convention of having the application in memory even
  // after all windows have been closed
  if (process.platform !== 'darwin') {
    app.quit();
  }
});

app.whenReady().then(createWindow).catch(console.log);

app.on('activate', () => {
  // On macOS it's common to re-create a window in the app when the
  // dock icon is clicked and there are no other windows open.
  if (mainWindow === null) createWindow();
});


/***
 * 
 */

 const { ipcMain } = require('electron');
 var recursive = require("recursive-readdir");
 const webp = require('webp-converter');
 webp.grant_permission();
 const fs = require('fs')
 
 var FolderLocation : string = null;
 let exclusionFolder = ["node_modules"];
 let quality = 80;
 
 ipcMain.on('ChooseFolder', () => {
   dialog.showOpenDialog({ properties: ['openDirectory'] }).then((resolve)=> {
    console.log(resolve.canceled);
     
    if(resolve.canceled === false) {
      FolderLocation = resolve.filePaths[0];
      let folderSelected : string = resolve.filePaths[0].replace(/\\/g, "/");    
      mainWindow.webContents.executeJavaScript(`
          document.getElementById("Selected").innerHTML = '${folderSelected}'
      `)
     }
   })
 });
 
 ipcMain.on('inputChanged', (channel, inputValue) => {
   quality = inputValue;
 });
 
const disableBtn  = (bool : boolean) => {
  console.log(bool);
  
  mainWindow.webContents.executeJavaScript(`
    document.getElementById("ChooseFolder").disabled = ${bool};
    document.getElementById("makeWebp").disabled = ${bool};
  `);
}

const setCurrentAction = (content : string) => {  
  console.log(content);
  
  mainWindow.webContents.executeJavaScript(`
    document.getElementById("CurrentAction").innerHTML = '${content}';
  `);
}

 ipcMain.on('startWebpGen', () => {
   if(FolderLocation !== null) {
    disableBtn(true);
    recursive(FolderLocation, [] , function (err : any, files : any) {      
      let rst = files.filter(function(file : any) {
          if (path.extname(file).toLowerCase() === ".png") return true;
          if (path.extname(file).toLowerCase() === ".jpg") return true;
          if (path.extname(file).toLowerCase() === ".jpeg") return true;
          return false;
      });
      let i = 0;
      setCurrentAction(`In progress ${i} / ${rst.length}`);

      const iterationFunction = ( ObjectArray : any) => {
        const result = webp.cwebp(ObjectArray[i], path.dirname(ObjectArray[i]) + "\\" + path.basename(ObjectArray[i], path.extname(ObjectArray[i]))  + ".webp","-q " + quality);
        result.then((result : any) => {
          ++i;
          if(i !== rst.length) { // diff de 3
            setCurrentAction(`In progress ${i} / ${rst.length}`);
            iterationFunction(ObjectArray);
          } else {
            console.log("finished"); 
            console.log("ici");
            setCurrentAction(`Ended successfully ${i} / ${rst.length}`);
            disableBtn(false);
          }
        });
      }; iterationFunction(rst)
      
    }); // End recursive
   } else {
      setCurrentAction(`Choose a folder`);
   }
 });